contrib: add support for confluent-kafka-go/v2/kafka

dferstay opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Commit and PR titles should be prefixed with the general area of the pull request's change.

-->
### What does this PR do?

Adds support for confluent-kafka-go/v2/kafka to the contrib repository.

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
* If this resolves a GitHub issue, include "Fixes #XXXX" to link the issue and auto-close it on merge.
-->
Fixes #1770 

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
* Ideally your changes have automated unit and/or integration tests which are run in CI.
-->

Unit tests included

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).
-->

- [ ] Changed code has unit tests for its functionality.
- [ ] If this interacts with the agent in a new way, a system test has been added.

dferstay commented:
@Julio-Guerra , @ahmed-mez , 
Would it be possible to get some feedback on this PR when you get a chance?

rarguelloF commented on contrib/confluentinc/confluent-kafka-go/kafka.v2/kafka.go:
```diff
@@ -0,0 +1,279 @@
+	"github.com/confluentinc/confluent-kafka-go/v2/kafka"
+)
+
+const componentName = "confluentinc/confluent-kafka-go/kafka"
```
```suggestion
const componentName = "confluentinc/confluent-kafka-go/kafka.v2"
```

rarguelloF commented on contrib/confluentinc/confluent-kafka-go/kafka.v2/kafka.go:
```diff
@@ -0,0 +1,279 @@
+	if out == nil {
+		return out
+	}
+
```
```suggestion
```

rarguelloF commented on contrib/confluentinc/confluent-kafka-go/kafka.v2/kafka.go:
```diff
@@ -0,0 +1,279 @@
+			span.Finish()
+		}
+	}()
+
```
```suggestion
```

rarguelloF commented on contrib/confluentinc/confluent-kafka-go/kafka.v2/kafka.go:
```diff
@@ -0,0 +1,279 @@
+	if spanctx, err := tracer.Extract(carrier); err == nil {
+		opts = append(opts, tracer.ChildOf(spanctx))
+	}
+
```
```suggestion
```

rarguelloF commented on contrib/confluentinc/confluent-kafka-go/kafka.v2/kafka.go:
```diff
@@ -0,0 +1,279 @@
+			oldDeliveryChan <- evt
+		}()
+	}
+
```
```suggestion
```

rarguelloF commented on contrib/confluentinc/confluent-kafka-go/kafka.v2/kafka.go:
```diff
@@ -0,0 +1,279 @@
+// Produce calls the underlying Producer.Produce and traces the request.
+func (p *Producer) Produce(msg *kafka.Message, deliveryChan chan kafka.Event) error {
+	span := p.startSpan(msg)
+
```
```suggestion
```

rarguelloF commented on contrib/confluentinc/confluent-kafka-go/kafka.v2/kafka.go:
```diff
@@ -0,0 +1,279 @@
+	if deliveryChan == nil {
+		span.Finish(tracer.WithError(err))
+	}
+
```
```suggestion
```

rarguelloF commented on contrib/confluentinc/confluent-kafka-go/kafka.v2/option.go:
```diff
@@ -0,0 +1,106 @@
+// WithContext sets the config context to ctx.
+// Deprecated: This is deprecated in favor of passing the context
+// via the message headers
+func WithContext(ctx context.Context) Option {
```
I don't think we need to add this deprecated option to the new integration, since people will have to update their code anyway. Any thoughts? @ajgajg1134 @ahmed-mez

rarguelloF replied:
@knusbaum

rarguelloF commented on contrib/confluentinc/confluent-kafka-go/kafka.v2/kafka.go:
```diff
@@ -0,0 +1,279 @@
+	if in == nil {
+		return nil
+	}
+
```
```suggestion
```

rarguelloF commented on contrib/confluentinc/confluent-kafka-go/kafka.v2/kafka.go:
```diff
@@ -0,0 +1,279 @@
+			if msg, ok := evt.(*kafka.Message); ok {
+				next = c.startSpan(msg)
+			}
+
```
```suggestion
```

rarguelloF commented on contrib/confluentinc/confluent-kafka-go/kafka.v2/kafka.go:
```diff
@@ -0,0 +1,279 @@
+			c.prev = nil
+		}
+	}()
+
```
```suggestion
```

rarguelloF commented on contrib/confluentinc/confluent-kafka-go/kafka.v2/example_test.go:
```diff
@@ -0,0 +1,86 @@
+
+// This example shows how a span context can be passed from a producer to a consumer.
+func Example() {
+
```
```suggestion
```

rarguelloF commented on contrib/confluentinc/confluent-kafka-go/kafka.v2/example_test.go:
```diff
@@ -0,0 +1,86 @@
+		"session.timeout.ms":       10,
+		"enable.auto.offset.store": false,
+	})
+
```
```suggestion
```

rarguelloF commented on contrib/confluentinc/confluent-kafka-go/kafka.v2/example_test.go:
```diff
@@ -0,0 +1,86 @@
+		tracer.Inject(parentSpan.Context(), carrier)
+
+		c.Consumer.Events() <- msg
+
```
```suggestion
```

rarguelloF commented on contrib/confluentinc/confluent-kafka-go/kafka.v2/example_test.go:
```diff
@@ -0,0 +1,86 @@
+	if err != nil {
+		panic(err)
+	}
+
```
```suggestion
```

rarguelloF reviewed (commented):
Overall looks good! Just a typo on the integration's component name and some whitespace removal as documented in our [style guidelines](https://github.com/DataDog/dd-trace-go/wiki/Style-guidelines):

> Stay compact. Avoid creating paragraphs by introducing empty lines as much as possible, especially around curly braces. The curly brace already creates enough spacing for a reader. Exceptions can be made in long functions where there is a big shift in purpose.

Also please take a look at the `go.mod` conflicts and the CI jobs that are failing.

dferstay commented:
> Overall looks good! Just a typo on the integration's component name and some whitespace removal as documented in our [style guidelines](https://github.com/DataDog/dd-trace-go/wiki/Style-guidelines):
> 
> > Stay compact. Avoid creating paragraphs by introducing empty lines as much as possible, especially around curly braces. The curly brace already creates enough spacing for a reader. Exceptions can be made in long functions where there is a big shift in purpose.
> 
> Also please take a look at the `go.mod` conflicts and the CI jobs that are failing.

@rarguelloF , thanks for the feedback!
I've made an attempt at addressing your comments and I think this PR is ready for another look. 👍

knusbaum commented on contrib/confluentinc/confluent-kafka-go/kafka.v2/example_test.go:
```diff
@@ -0,0 +1,82 @@
+
+	kafkatrace "gopkg.in/DataDog/dd-trace-go.v1/contrib/confluentinc/confluent-kafka-go/kafka.v2"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+)
```
Nit:
`"github.com/confluentinc/confluent-kafka-go/v2/kafka"` should come last here.

Stdlib
<line>
Internal (same repo)
<line>
External

dferstay replied:
ACK - rearranged.

knusbaum reviewed (changes requested):
nitpick, otherwise looks good.

dferstay commented:
Thanks for merging `main` into my branch, @rarguelloF .

It seems that the tests for `contrib/google.golang.org/api` are failing; will this be fixed by https://github.com/DataDog/dd-trace-go/pull/1939 ?

dferstay commented:
@rarguelloF , @knusbaum , 
I've merged main with this branch which pulls in https://github.com/DataDog/dd-trace-go/pull/1937 .  Is this change good to go?

rarguelloF commented:
@dferstay 👋 
https://github.com/DataDog/dd-trace-go/pull/1971 has been merged, so we can try again to update this branch and see if the CI issue is resolved. 

I went ahead and resolved conflicts in the `go.mod` and `go.sum` files, so if CI passes it should be good to merge 👍

dferstay commented:
> @dferstay 👋 #1971 has been merged, so we can try again to update this branch and see if the CI issue is resolved.
> 
> I went ahead and resolved conflicts in the `go.mod` and `go.sum` files, so if CI passes it should be good to merge 👍

Thank you; we're green!
